File: ngwart/ngwart/engine/events.py
"""Engine -> UI event channel.

The engine never imports Qt. It emits plain dataclass events to a listener,
and the Qt layer adapts those into signals (see ui/bridge.py). That keeps the
whole sequencer testable headlessly -- the test suite runs it with a recording
listener and asserts on the event stream.

It also means the engine thread never touches a widget, which is the actual
root cause of v1's frozen UI: there, the Tk update loop and the test loop were
coroutines on one event loop, so a blocking serial read froze the display.
"""

from __future__ import annotations

import threading
from dataclasses import dataclass, field
from typing import Any, Protocol


# --- events -------------------------------------------------------------

@dataclass
class LogEvent:
    message: str
    level: str = "info"  # info | warn | error | pass | fail
    row: int | None = None


@dataclass
class StepEvent:
    row: int
    text: str
    comment: str = ""


@dataclass
class StatusEvent:
    text: str
    color: str | None = None


@dataclass
class ProgressEvent:
    #: 0.0-1.0, or -1.0 for indeterminate.
    value: float


@dataclass
class TimerEvent:
    seconds: float


@dataclass
class GridEvent:
    """A mutation of one of the result grids."""

    grid: int
    op: str                      # add | clear | place | unplace | config
    values: list[str] = field(default_factory=list)
    tag: str = ""
    config: dict = field(default_factory=dict)


@dataclass
class FieldEvent:
    """A named single-value UI field (barcode entries, counters, monitors)."""

    name: str
    value: str
    color: str | None = None


@dataclass
class AliveEvent:
    alive: list[int]


@dataclass
class RunStateEvent:
    state: str                   # idle | validating | running | stopping | teardown
    detail: str = ""


@dataclass
class ResultEvent:
    """Terminal verdict for the run."""

    passed: bool
    per_uut: dict[int, bool] = field(default_factory=dict)
    detail: str = ""


Event = Any


# --- listeners ----------------------------------------------------------

class Listener(Protocol):
    def emit(self, event: Event) -> None: ...


class NullListener:
    """Discards everything. Default so the engine can run truly headless."""

    def emit(self, event: Event) -> None:  # noqa: D102
        pass


class RecordingListener:
    """Keeps every event. Used by the test suite and the run recorder."""

    def __init__(self) -> None:
        self.events: list[Event] = []
        self._lock = threading.Lock()

    def emit(self, event: Event) -> None:
        with self._lock:
            self.events.append(event)

    def of_type(self, cls) -> list:
        with self._lock:
            return [e for e in self.events if isinstance(e, cls)]

    def messages(self) -> list[str]:
        return [e.message for e in self.of_type(LogEvent)]

    def clear(self) -> None:
        with self._lock:
            self.events.clear()


class FanOut:
    """Broadcasts to several listeners.

    Registration is guarded because the UI may attach while the engine thread
    is already emitting.
    """

    def __init__(self, *listeners: Listener) -> None:
        self._listeners: list[Listener] = list(listeners)
        self._lock = threading.Lock()

    def add(self, listener: Listener) -> None:
        with self._lock:
            self._listeners.append(listener)

    def remove(self, listener: Listener) -> None:
        with self._lock:
            if listener in self._listeners:
                self._listeners.remove(listener)

    def emit(self, event: Event) -> None:
        with self._lock:
            targets = list(self._listeners)
        for listener in targets:
            # A misbehaving listener must never take down a test run mid-fixture.
            try:
                listener.emit(event)
            except Exception:  # noqa: BLE001
                pass

File: ngeoltx/src/Ngeoltx.Engine/Events.cs
namespace Ngeoltx.Engine;

/// <summary>
/// Engine to UI channel.
/// </summary>
/// <remarks>
/// The engine never references a UI framework. It publishes records to an
/// <see cref="IEngineListener"/>, and the Avalonia layer adapts those onto the
/// dispatcher. That keeps the sequencer testable headlessly, and means the
/// engine thread never touches a control -- the actual root cause of v1's
/// frozen UI, where the Tk update loop and the test loop shared one event loop.
/// </remarks>
public interface IEngineEvent { }

public sealed record LogEvent(string Message, string Level = "info", int? Row = null)
    : IEngineEvent;

public sealed record StepEvent(int Row, string Text, string Comment = "")
    : IEngineEvent;

public sealed record StatusEvent(string Text, string? Colour = null) : IEngineEvent;

/// <param name="Value">0..1, or -1 for indeterminate.</param>
public sealed record ProgressEvent(double Value) : IEngineEvent;

public sealed record TimerEvent(double Seconds) : IEngineEvent;

/// <summary>A mutation of one result grid.</summary>
public sealed record GridEvent(
    int Grid,
    string Op,                                  // add | clear | place | config
    IReadOnlyList<string>? Values = null,
    string Tag = "",
    IReadOnlyDictionary<string, object>? Config = null) : IEngineEvent;

/// <summary>A named single-value UI field: barcode entries, counters.</summary>
public sealed record FieldEvent(string Name, string Value, string? Colour = null)
    : IEngineEvent;

public sealed record AliveEvent(IReadOnlyList<int> Alive) : IEngineEvent;

/// <param name="State">idle | validating | running | stopping | teardown</param>
public sealed record RunStateEvent(string State, string Detail = "") : IEngineEvent;

public sealed record ResultEvent(
    bool Passed,
    IReadOnlyDictionary<int, bool> PerUut,
    string Detail = "") : IEngineEvent;


public interface IEngineListener
{
    void Emit(IEngineEvent e);
}

/// <summary>Discards everything, so the engine can run truly headless.</summary>
public sealed class NullListener : IEngineListener
{
    public static readonly NullListener Instance = new();
    public void Emit(IEngineEvent e) { }
}

/// <summary>Keeps every event. Used by tests and the run recorder.</summary>
public sealed class RecordingListener : IEngineListener
{
    private readonly List<IEngineEvent> _events = new();
    private readonly Lock _gate = new();

    public void Emit(IEngineEvent e)
    {
        lock (_gate) _events.Add(e);
    }

    public IReadOnlyList<IEngineEvent> Events
    {
        get { lock (_gate) return _events.ToList(); }
    }

    public IEnumerable<T> OfType<T>() where T : IEngineEvent =>
        Events.OfType<T>();

    public void Clear()
    {
        lock (_gate) _events.Clear();
    }
}

/// <summary>Broadcasts to several listeners.</summary>
public sealed class FanOut : IEngineListener
{
    private readonly List<IEngineListener> _listeners;
    private readonly Lock _gate = new();

    public FanOut(params IEngineListener[] listeners) => _listeners = listeners.ToList();

    public void Add(IEngineListener listener)
    {
        lock (_gate) _listeners.Add(listener);
    }

    public void Emit(IEngineEvent e)
    {
        List<IEngineListener> targets;
        lock (_gate) targets = _listeners.ToList();
        foreach (var listener in targets)
        {
            // A misbehaving listener must never take down a run mid-fixture.
            try { listener.Emit(e); }
            catch { /* ignored on purpose */ }
        }
    }
}
